### PR TITLE
Deprecate `shoot.spec.dns.providers` handling

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area cleanup
/kind api-change

**What this PR does / why we need it**:
This PRs adds deprecation notes about the DNS provider handling. See https://github.com/gardener/gardener/issues/9176 for more information.

**Which issue(s) this PR fixes**:
Part of #9176

**Special notes for your reviewer**:
/cc @MartinWeindel

**Release note**:

```breaking user
The specification of additional, non-primary DNS providers was deprecated and will be discontinued in a future release. If you need additional DNS providers for your shoot workload, please use the provider config for the respective DNS extension.
```

### CODE REVIEW

File: example/90-shoot.yaml [hunk 1]
```diff
@@ -314,10 +314,10 @@ spec:
     # When the shoot shall use a cluster domain no domain and no providers need to be provided - Gardener will
     # automatically compute a correct domain based on the default domains in the garden cluster.
     domain: crazy-botany.core.my-custom-domain.com
+    # Provider configuration required if custom shoot domain is configured.
   # providers:
   # - type: aws-route53
   #   secretName: my-custom-domain-secret
-  #   primary: true # `true` indicates that this provider is also used to manage the shoot domain `.spec.dns.domain`
```
Comment (MartinWeindel) [inline example/90-shoot.yaml:320]: Is it not too early to remove the primary field? You still need to set it.

File: pkg/apis/core/types_shoot.go [hunk 2]
```diff
@@ -351,17 +351,19 @@ type DNS struct {
 	Domain *string
 	// Providers is a list of DNS providers that shall be enabled for this shoot cluster. Only relevant if
 	// not a default domain is used.
+	// Deprecated: Configuring multiple DNS providers is deprecated and will be forbidden in a future release. Please use the DNS extension provider config for additional providers.
```
Comment (MartinWeindel) [inline pkg/apis/core/types_shoot.go]: Better use the full name of the shoot-dns-service extension.
```suggestion
	// Deprecated: Configuring multiple DNS providers is deprecated and will be forbidden in a future release. Please use the `shoot-dns-service` extension provider config for additional providers.
```

File: pkg/apis/core/types_shoot.go [hunk 3]
```diff
@@ -351,17 +351,19 @@ type DNS struct {
...
 
+// TODO(timuthy): Rework the 'DNSProvider' struct and deprecated fields in the scope of https://github.com/gardener/gardener/issues/9176.
+
 // DNSProvider contains information about a DNS provider.
 type DNSProvider struct {
-	// TODO(timuthy): Remove this field in the scope of https://github.com/gardener/gardener/issues/9176.
-
 	// Domains contains information about which domains shall be included/excluded for this provider.
-	// Deprecated: This field is deprecated and will be removed in a future release.
+	// Deprecated: This field is deprecated and will be removed in a future release. Please use the DNS extension provider config for additional configuration.
```
Comment (MartinWeindel) [inline pkg/apis/core/types_shoot.go]:
```suggestion
	// Deprecated: This field is deprecated and will be removed in a future release. Please use the `shoot-dns-service` extension provider config for additional configuration.
```

File: pkg/apis/core/types_shoot.go [hunk 4]
```diff
@@ -371,10 +373,8 @@ type DNSProvider struct {
 	// Type is the DNS provider type for the Shoot. Only relevant if not the default domain is used for
 	// this shoot.
 	Type *string
-	// TODO(timuthy): Remove this field in the scope of https://github.com/gardener/gardener/issues/9176.
-
 	// Zones contains information about which hosted zones shall be included/excluded for this provider.
-	// Deprecated: This field is deprecated and will be removed in a future release.
+	// Deprecated: This field is deprecated and will be removed in a future release. Please use the DNS extension provider config for additional configuration.
```
Comment (MartinWeindel) [inline pkg/apis/core/types_shoot.go]:
```suggestion
	// Deprecated: This field is deprecated and will be removed in a future release. Please use the `shoot-dns-service` extension provider config for additional configuration.
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 5]
```diff
@@ -432,21 +432,23 @@ type DNS struct {
 	Domain *string `json:"domain,omitempty" protobuf:"bytes,1,opt,name=domain"`
 	// Providers is a list of DNS providers that shall be enabled for this shoot cluster. Only relevant if
 	// not a default domain is used.
+	// Deprecated: Configuring multiple DNS providers is deprecated and will be forbidden in a future release. Please use the DNS extension provider config for additional providers.
```
Comment (MartinWeindel) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// Deprecated: Configuring multiple DNS providers is deprecated and will be forbidden in a future release. Please use the `shoot-dns-service` extension provider config for additional providers.
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 6]
```diff
@@ -432,21 +432,23 @@ type DNS struct {
...
 
+// TODO(timuthy): Rework the 'DNSProvider' struct and deprecated fields in the scope of https://github.com/gardener/gardener/issues/9176.
+
 // DNSProvider contains information about a DNS provider.
 type DNSProvider struct {
-	// TODO(timuthy): Remove this field in the scope of https://github.com/gardener/gardener/issues/9176.
-
 	// Domains contains information about which domains shall be included/excluded for this provider.
-	// Deprecated: This field is deprecated and will be removed in a future release.
+	// Deprecated: This field is deprecated and will be removed in a future release. Please use the DNS extension provider config for additional configuration.
```
Comment (MartinWeindel) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// Deprecated: This field is deprecated and will be removed in a future release. Please use the `shoot-dns-service` extension provider config for additional configuration.
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 7]
```diff
@@ -458,10 +460,9 @@ type DNSProvider struct {
 	// Type is the DNS provider type.
 	// +optional
 	Type *string `json:"type,omitempty" protobuf:"bytes,4,opt,name=type"`
-	// TODO(timuthy): Remove this field in the scope of https://github.com/gardener/gardener/issues/9176.
 
 	// Zones contains information about which hosted zones shall be included/excluded for this provider.
-	// Deprecated: This field is deprecated and will be removed in a future release.
+	// Deprecated: This field is deprecated and will be removed in a future release. Please use the DNS extension provider config for additional configuration.
```
Comment (MartinWeindel) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// Deprecated: This field is deprecated and will be removed in a future release. Please use the `shoot-dns-service` extension provider config for additional configuration.
```

**Review by MartinWeindel — CHANGES_REQUESTED** [review]
Thanks for improving the deprecated messages. It would be clearer to use the full name of the shoot-dns-service extension.

File: example/90-shoot.yaml (hunk 1 above)
Comment (timuthy) [reply example/90-shoot.yaml:320]: It is set/defaulted by Gardener if not set. The removal from the example was more or less a simplification and encouragement to not set it.

File: pkg/apis/core/types_shoot.go (hunk 2 above)
Comment (timuthy) [reply pkg/apis/core/types_shoot.go]: How can we know that it's the `shoot-dns-service`?

Comment (MartinWeindel) [reply pkg/apis/core/types_shoot.go]: I understand the intention to be neutral. As there is no known other DNS extension listed on [Known extension implementations list](https://github.com/gardener/gardener/blob/b26920f63445597d107fccadd05a19cc5783ca2c/extensions/README.md#known-extension-implementations), I don't see an advantage not to named it.
But it was only a suggestion.

**Review by MartinWeindel — APPROVED** [review]
lgtm

File: pkg/apis/core/types_shoot.go (hunk 2 above)
Comment (timuthy) [reply pkg/apis/core/types_shoot.go]: Okay, thanks. Pease see the adjusted doc string.

**Review by MartinWeindel — APPROVED** [review]
lgtm

**Review by rfranzke — COMMENTED** [review]
/approve

**Pull request merged**